### DISCUSSION TITLE
Graph Explorer

### DISCUSSION THREAD

**Discussion by Abhay22** [JanusGraph/janusgraph#2292] [Ideas]
Graph Explorer project is very much inactive so, I was thinking if we can fork that project and keep under JanusGraph then there will be a lot of contributions.
I made a lot of changes while working on it but as GraphExplorer is inactive so, it is waste raising any PR there.
For me Security was main concern so I moved from directly UI remotely connecting on WS to UI connecting to a spring boot app which we can secure as per our need. (Here I tried many options to connect on WSS but none of the solution worked for e)

- Single Page UI with Springboot app for backed support
- Protocol: Websocket, REST & REST ON JAVA ( Websocket & REST are default options and I introduced other option of REST )
- Results Limit: 500 Nodes (As we have to deal with huge data and it may lead to timeout if we don't put any limit )
- Add/Update Graph Support : No (I kept is only readonly and I don't want anybody to edit graph using a visualizer but still we can keep if required)
- Drop Query Supported: No (No one should drop any data using visualizer but still we can keep if required)
- Item Info scrollable bug: Fixed
- Nodes & Edges are less responsive : Fixed
- Nodes disappears after 5 clicks : It will just fade inactive nodes but will not disappear completely
- Application server required : Yes 
- Custom Query : We can write any custom query using for just getting nodes.
- Default open source layout: It supports default layout with protocol "Websocket" and "Rest" 
- Looks: I have given it a look of Target colour with Target logo
- Data Control: Yes (Now we have more control on data. Which data we want to show and which not it will be controlled in backend)
- Data Input Type: Janusgraph and files with nodes and edges information.Single Page UI with Springboot app for backed support
- Protocol: Websocket, REST & REST ON JAVA ( Websocket & REST are default options and I introduced other option of REST )
- Results Limit: 500 Nodes (As we have to deal with huge data and it may lead to timeout if we don't put any limit )
- Item Info scrollable bug: Fixed
- Nodes & Edges are less responsive : Fixed
- Nodes disappears after 5 clicks : It will just fade inactive nodes but will not disappear completely
- Custom Query : We can write any custom query using for just getting nodes.
- Default open source layout: It supports default layout with protocol "Websocket" and "Rest" but only for local usage
- Data Control: Yes (Now we have more control on data. Which data we want to show and which not it will be controlled in backend)
- Data Input Type: Janusgraph and files with nodes and edges information. (need some changes if required)

Please let me know your thoughts.

**Comment by li-boxuan** [comment]
Which graph explorer project you are talking about? Is it https://github.com/invanalabs/graph-explorer? This one seems active though.

**Reply by Abhay22** [reply]
> Which graph explorer project you are talking about? Is it https://github.com/invanalabs/graph-explorer? This one seems active though.

I am referring to this: https://github.com/bricaud/graphexp

**Reply by li-boxuan** [reply]
Looks like it supports all Tinkerpop compatible graph databases. Thus, I don't think it makes much sense to fork it under the JanusGraph namespace. I also doubt if we have enough human resources to review/maintain an additional frontend project. I would suggest you put it under your own namespace, and probably we can put a link to your fork in README. It's just my personal view though.

**Reply by FlorianHockmann** [reply]
It may also be worthwhile to send the author @bricaud a message to see whether he would welcome such PRs and have the time to review them or whether there's another option to collaborate before just forking his project.
But I agree with @li-boxuan in general that we shouldn't fork an external UI project into the JanusGraph organization. We currently wouldn't have the manpower to maintain it any better than the original author(s) and we could lose possible contributions from users of other graph DBs. So, if you want to fork it, then you would have to do that outside of the JanusGraph org. But again, I wouldn't fork a project if there's a chance to collaborate with the existing authors.

**Reply by bricaud** [reply]
Thanks for mentioning me! I do not have much time for graphexp but if I get PRs I usually merge them, if the changes are not too complex and not too messy.
I would be happy if graphexp is developed further and useful for someone, so if you see anything that would make people contribute more, let me know. If it implies moving the code out of my private repo to a new git repo where several people can be admin, I would be happy to do so.
For the suggestion about forking into Janusgraph, I agree, application to other graph DBs would be lost. So what about moving it somewhere else? on his own maybe?
There is another contributor @armandleopold who focus on the UI, he may contribute as well.
